grammar/rainerscript.c

```diff
@@ -2122,6 +2122,50 @@ doFuncCall(struct cnffunc *__restrict__ const func, struct svar *__restrict__ co
 		if(bMustFree) free(str);
 		varFreeMembers(&r[1]);
 		break;
+	case CNFFUNC_FORMAT_TIME: {
+		time_t unixtime;
+		struct tm lt;
+		char   result[64];
+		char  *formatstr = NULL, *endptr = NULL;
+
+		cnfexprEval(func->expr[0], &r[0], usrptr);
+		cnfexprEval(func->expr[1], &r[1], usrptr);
+
+		str = (char*) var2CString(&r[0], &bMustFree);
+		formatstr = (char*) es_str2cstr(r[1].d.estr, NULL);
+
+
+		// TODO: Error checking, and account for different
+		//       possible sizes of time_t 32/64 bit.
+		errno = 0;
+		unixtime = (time_t) strtoll(str, &endptr, 10);
+
+		printf("STR: %s, FORMATSTR: %s, NUM: %ld\n", str, formatstr, unixtime);
+
+		localtime_r(&unixtime, &lt);
+
+		if (strncmp(formatstr, "rfc3164", 7) == 0 || strncmp(formatstr, "rfc3164", 7) == 0) {
+			strftime(result, 64, "%b %e %H:%M:%S", &lt);
+		} else if (strncmp(formatstr, "rfc3339", 7) == 0 || strncmp(formatstr, "RFC3339", 7) == 0) {
+			strftime(result, 64, "%Y-%m-%dT%H:%M:%S%z", &lt);
```

Comment:
Same here: I am a bit concerned on the use of strftime. The man page says it consults TZ and LC_TIME. However, the timestamp may have orginated in many different timezones. Are we sure the result is what we want?

Reply:
I did some quick testing on Linux (I'll do more), and I think the man page for `strftime()` is just misleading (no surprise there).

The glibc version of `strftime()` doesn't actually manipulate the time/date it's passed before producing its output. In fact, according to a comment in a recent copy of `strftime_l.c`:

```c
/* The POSIX test suite assumes that setting
   the environment variable TZ to a new value before calling strftime()
   will influence the result (the %Z format) even if the information in
   TP is computed with a totally different time zone.
   This is bogus: though POSIX allows bad behavior like this,
   POSIX does not require it.  Do the right thing instead.  */
```
So, at least in  GNU, it has no effect. `strftime()` itself _does_ indeed call `tzset()`, but it's only to set a flag, and to satisfy POSIX. 

I've tested this with the following simple program:

```c
#include <time.h>
#include <stdio.h>

int main(void) {

   time_t t = 0xdeadbeef;
   struct tm tm;

   int max = 64;
   char result[max];

   localtime_r(&t, &tm);


   strftime(result, max, "%Y-%m-%d %H:%M:%S %z", &tm);
   printf("RESULT: %s\n", result);
   printf("DST?: %d\n", tm.tm_isdst);

   tzset();
   gmtime_r(&t, &tm);

   strftime(result, max, "%Y-%m-%d %H:%M:%S %z", &tm);
   printf("RESULT: %s\n", result);
   printf("DST?: %d\n", tm.tm_isdst);

   return 0;
}
```

And, indeed, changing `TZ` has no effect on the time produced by `gmtime()`/`strftime()` (including `%Z`).

That's GNU. The C library on Solaris seems to be a bit of a different story (according to the latest CIs). The Solaris man 
 page `strftime()` doesn't mention `TZ` like the GNU one does. Solaris seems to respect the same rules as far as ignoring the `TZ` value with respect to `%Z`, but for some reason it preserves the `tm_isdst` flag of `struct tm` resulting in UTC/GMT times with a `+0100` value for `%Z`. 🙄 

I think that can be addressed by simply setting `tm_isdst` to zero prior to calling `strftime()` (I still have to test this). This should be safe as the test output for Solaris matches the output for GNU, as far as the date/time parameters are concerned -- as in, the dates and times are the same, it's just the `+01:00` that's different.

<hr />

Since UNIX timestamps are technically supposed to be expressed in terms of UTC, then `format_time()` should probably represent them this way (when `date-rfc3339` or `%Z` are used). Perhaps later versions could allow the user to manipulate the `%Z` value to match the actual timezone of the remote log source. What do you think?

Reply:
Great analysis!

And: yeah, I was always very skeptic of strftime(). Actually the portability problems are one reason (I now remember in more detail) of staying away from it. There may even be such a discussion here or in the mailing list archive.

That said, I think it would be best to implement our own version of strftime(), at least for the common formats. Otherwise, we'll probably end up with the need to implement it anyhow. This is not uncommon to happen years later, which means usually I am personally than in trouble -- thus I have begun to point out these things very early.

Reply:
Oh, and I forgot: testing these things never gives the definite answer. Does, for example, AIX handle it the same way? How about BSD? What in regard to old versions of Solaris? And ultimately: what about upcoming OS vendor patches?

That's also the reason why rsyslog has it's own rs_strerror() which works around all the bad things you find in the distros. Unfortunately, I don't think it is so easy to do with strftime() as well.